File: use-cases/pricing/code/python/src/calculate_new_price_handler.py
import logging
from src.utils.pricing_utils import PricingOfferLambdaInput, Money
from src.utils import constants

# Set up logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)


# Lambda handler function
def lambda_handler(event, context):
    # Log the input event for the Lambda
    logger.info(f"CalculateNewPrice Lambda input {event}")

    try:
        # Instantiate PricingOfferLambdaInput object from the input event
        pricing_offer = PricingOfferLambdaInput(**event)

        # Calculate the landed price by summing listing price and shipping price
        landed_price = pricing_offer.sellerOffer.listingPrice.amount + pricing_offer.sellerOffer.shippingPrice.amount

        new_item_price = None

        # Condition to check if competitivePriceThreshold is present. Pricing Health flow
        if pricing_offer.sellerOffer.referencePrice and pricing_offer.useCompetitivePrice:

            new_item_price = pricing_offer.sellerOffer.referencePrice["competitivePriceThreshold"][
                "amount"]
            result = calculate_price(new_item_price, pricing_offer,
                                     pricing_offer.sellerOffer.referencePrice["competitivePriceThreshold"][
                                         "currency_code"])

            logger.info(f"new listing price: {new_item_price}")

        # Check conditions to determine whether to skip new price calculation
        # Check if buy box price is less than the minimum threshold
        elif pricing_offer.buyBox.price.amount < pricing_offer.minThreshold:
            # Log and return indicating skipping new price calculation
            logger.info(
                f"Buy Box Price: {pricing_offer.buyBox.price.amount} is less than threshold: "
                f"{pricing_offer.minThreshold}. "
                f"- Skipping new price calculation.")
            return {
                "newListingPrice": {
                    "amount": -1
                },
                "issues": f"Buy Box Price: {pricing_offer.buyBox.price.amount} is less than threshold"
            }

        # Check if buy box price is greater than landed price
        elif pricing_offer.buyBox.price.amount > landed_price:
            # Log and return indicating skipping new price calculation
            logger.info(
                f"Landed Price: {landed_price} is already less than Buy Box Price: "
                f"{pricing_offer.buyBox.price.amount}. "
                f"- Skipping new price calculation.")
            return {
                "newListingPrice": {
                    "amount": -1
                },
                "issues": f"Landed Price: {landed_price} is already less than Buy Box Price"
            }

        # Calculate the new item price based on different price change rules (percentage or fixed)
        elif pricing_offer.priceChangeRule.rule == constants.PriceChangeRule.PERCENTAGE.value:
            new_item_price = (pricing_offer.buyBox.price.amount - pricing_offer.sellerOffer.shippingPrice.amount - (
                    pricing_offer.priceChangeRule.value / 100)
                              * pricing_offer.buyBox.price.amount)
        elif pricing_offer.priceChangeRule.rule == constants.PriceChangeRule.FIXED.value:
            new_item_price = pricing_offer.buyBox.price.amount - pricing_offer.priceChangeRule.value - pricing_offer.sellerOffer.shippingPrice.amount
        else:
            # Log and return indicating invalid price change rule
            logger.info(
                f"Price Change Rule: {pricing_offer.priceChangeRule.rule} is Invalid."
                f"- Skipping new price calculation."
                f"- Please change rule to match one of [PERCENTAGE, FIXED]")
            return {
                "newListingPrice": {
                    "amount": -1
                },
                "issues": f"Price Change Rule: {pricing_offer.priceChangeRule.rule} is Invalid."
            }

        # Assign new listing price to the calculate item price
        new_listing_price = new_item_price
        result = calculate_price(new_listing_price, pricing_offer, pricing_offer.sellerOffer.listingPrice.currency_code)

        return result
    except Exception as e:
        # Raise an exception if there's an error during the process
        raise Exception("Calling Pricing API failed", e)


def calculate_price(new_listing_price, pricing_offer, currency_code):
    # Check if the new listing price is less than the minimum threshold
    if new_listing_price < pricing_offer.minThreshold:
        # Log and return indicating skipping new price calculation
        logger.info(
            f"New Listings Price: {new_listing_price} is less than threshold: "
            f"{pricing_offer.minThreshold}. "
            f"- Skipping new price calculation.")
        return {
            "newListingPrice": {
                "amount": -1
            },
            "issues": f"Buy Box Price: {pricing_offer.buyBox.price.amount} is less than threshold"
        }

    # Create a Money object with the new listing price and currency code
    new_listing_price = Money(amount=new_listing_price,
                              currency_code=currency_code)
    # Prepare the result dictionary with the new listing price
    result = {
        "newListingPrice": new_listing_price.__dict__
    }
    return result

File: use-cases/pricing/code/java/src/main/java/lambda/CalculateNewPriceHandler.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.google.gson.Gson;
import lambda.utils.Amount;
import lambda.utils.Offer;
import lambda.utils.PriceChangeRule;
import lambda.utils.PricingLambdaInput;

import java.math.BigDecimal;

import static lambda.utils.Constants.PRICE_CHANGE_RULE_FIXED;
import static lambda.utils.Constants.PRICE_CHANGE_RULE_PERCENTAGE;

public class CalculateNewPriceHandler implements RequestHandler<PricingLambdaInput, PricingLambdaInput> {

    public PricingLambdaInput handleRequest(PricingLambdaInput input, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("CalculateNewPrice Lambda input: " + new Gson().toJson(input));

        try {
            //Calculate the landed price by summing listing price and shipping price
            Offer sellerOffer = input.getSellerOffer();
            float landedPrice = sellerOffer.getListingPrice().getAmount() + sellerOffer.getShippingPrice().getAmount();

            float buyBoxPrice = input.getBuyBox().getPrice().getAmount();

            //Check conditions to determine whether to skip new price calculation

            //Pricing Health Workflow - Check whether competitivePriceThreshold is present
            if (sellerOffer.referencePrice != null && input.useCompetitivePrice) {
                float newItemPrice = sellerOffer.referencePrice.getCompetitivePriceThreshold().getAmount().floatValue();
                return createNewPrice(newItemPrice, input.getMinThreshold(), buyBoxPrice, sellerOffer.referencePrice.getCompetitivePriceThreshold().getCurrencyCode(), logger);
            }

            //Check if buy box price is less than the minimum threshold
            if (buyBoxPrice < input.getMinThreshold()) {
                //Log and return indicating skipping new price calculation
                logger.log(String.format("Buy Box Price: %f is less than threshold: %f. Skipping new price calculation.",
                        buyBoxPrice,
                        input.getMinThreshold()));

                return PricingLambdaInput.builder()
                        .newListingPrice(Amount.builder()
                                .amount(-1)
                                .build())
                        .issues(String.format("Buy Box Price: %s is less than threshold", buyBoxPrice))
                        .build();
            }

            //Check if buy box price is greater than landed price
            if (buyBoxPrice > landedPrice) {
                logger.log(String.format("Landed Price: %f is already less than Buy Box Price: %f. Skipping new price calculation.",
                        landedPrice,
                        buyBoxPrice));

                return PricingLambdaInput.builder()
                        .newListingPrice(Amount.builder()
                                .amount(-1)
                                .build())
                        .issues(String.format("Landed Price: %f is already less than Buy Box Price", landedPrice))
                        .build();
            }

            //Calculate the new item price based on different price change rules (percentage or fixed)
            PriceChangeRule priceChangeRule = input.getPriceChangeRule();
            float newItemPrice;
            float buyBoxPriceExcludingShipping = buyBoxPrice - sellerOffer.getShippingPrice().getAmount();
            if (PRICE_CHANGE_RULE_PERCENTAGE.equals(priceChangeRule.getRule())) {
                newItemPrice = subtractPercentage(buyBoxPriceExcludingShipping, priceChangeRule.getValue());
            } else if (PRICE_CHANGE_RULE_FIXED.equals(priceChangeRule.getRule())) {
                newItemPrice = subtractFixed(buyBoxPriceExcludingShipping, priceChangeRule.getValue());
            } else {
                newItemPrice = -1;
                logger.log(String.format("Price Change Rule: %s is Invalid. Skipping new price calculation." +
                                "Please change rule to match one of [PERCENTAGE, FIXED]",
                        priceChangeRule.getRule()));

                return PricingLambdaInput.builder()
                        .newListingPrice(Amount.builder()
                                .amount(newItemPrice)
                                .build())
                        .issues(String.format("Price Change Rule: %s is Invalid.", priceChangeRule.getRule()))
                        .build();
            }

            //Calculate the new listing price by subtracting shipping price from the new item price
            float newListingPrice = newItemPrice;
            return createNewPrice(newListingPrice, input.getMinThreshold(), buyBoxPrice, input.sellerOffer.getListingPrice().currencyCode, logger);
        } catch (Exception e) {
            throw new InternalError("CalculateNewPrice Lambda failed", e);
        }
    }

    private PricingLambdaInput createNewPrice(float newListingPrice, float minThreshold, float buyBoxPrice, String offerCurrency, LambdaLogger logger) {
        //Check if the new listing price is less than the minimum threshold
        PricingLambdaInput.PricingLambdaInputBuilder pricingLambdaBuilder = PricingLambdaInput.builder();
        if (newListingPrice < minThreshold) {
            logger.log(String.format("New Listings Price: %f is less than threshold: %f. Skipping new price calculation.",
                    newListingPrice,
                    minThreshold));

            pricingLambdaBuilder
                    .newListingPrice(Amount.builder()
                            .amount(-1)
                            .build())
                    .issues(String.format("Buy Box Price: %f is less than threshold", buyBoxPrice));
        } else {
            pricingLambdaBuilder.newListingPrice(Amount.builder()
                            .currencyCode(offerCurrency)
                            .amount(newListingPrice)
                            .build())
                    .build();
        }
        return pricingLambdaBuilder.build();
    }

    private float subtractPercentage(float n1, float percentage) {
        return BigDecimal.valueOf(n1)
                .subtract(BigDecimal.valueOf(n1)
                        .multiply(BigDecimal.valueOf(percentage)))
                .floatValue();
    }

    private float subtractFixed(float n1, float n2) {
        return BigDecimal.valueOf(n1)
                .subtract(BigDecimal.valueOf(n2))
                .floatValue();
    }
}
